### ISSUE TITLE
quark wslib might be not optimized 

### ISSUE THREAD

**Issue by qcmv22** [supercollider/supercollider#4670]
Why is it so long to generate a midi File from data with the wslib quark.  I did a big try with
22,000 notes for an AI project  and after 1 hour I am still waiting !!!  it's more time then training a Neural Network for 2, 3  epochs.

Something  wrong, not optimized,  it's a sorting data problem...

(
 m = SimpleMIDIFile( "~/Desktop/LondonMidi_10dec_4track_T60.mid") ;
m.init1(5, 60, "4/4" );   //  4 tracks

m.addNote(72,  2, 2.45, 0.12, 127, track: 1 ) ;
m.addNote(80, 2, 2.65, 0.25, 127, track: 1 ) ;
m.addNote(73, 2, 2.76, 0.19, 127, track: 1 ) ;
m.addNote(50, 2, 3.17, 0.36, 127, track: 1 ) ;
m.addNote(43, 2, 4.18, 0.55, 127, track: 1 ) ;
(...)

It's seem ok just for very small project ( 100 to 1000 midi notes) .

Or maybe the default setting  in the interpreter cannot handle too much data... 

Best Regards

**Comment by joshpar** [comment]
the main SuperCollider project can not support Quark issues. Please file an issue in the project for this quark. Thanks!

**Comment by joshpar** [comment]
see https://github.com/supercollider-quarks/quarks

**Comment by qcmv22** [comment]
Wow you're faster then  the quark..

**Comment by jamshark70** [comment]
```
addMIDIEvent { |event, sort = true| ... }
```

So you should do `m.addMIDIEvent([...], false)` to add without sorting, and then *one* time at the end, call `sortMIDIEvents`.

Just stating this as a courtesy -- Josh is correct that this is not a problem with the SuperCollider core distribution (and, in fact, not even a bug in the extension class, because it already provides a way to do what you want). Further questions should be addressed to the user forum scsynth.org or mailing list.

**Comment by qcmv22** [comment]
Thanks James and Joshua... ok it's better with addMIDIEvent(...)  I realize that sorting it's not a linear process.  I did the low notes first  4300 notes : take 2 minutes  . Then I didthe higher notes  9660  : it took  12 minutes.  

Best.
    Le mardi 10 décembre 2019 19:32:02 UTC−5, H. James Harkins <notifications@github.com> a écrit :  
 
 addMIDIEvent { |event, sort = true| ... }

So you should do m.addMIDIEvent([...], false) to add without sorting, and then one time at the end, call sortMIDIEvents.

Just stating this as a courtesy -- Josh is correct that this is not a problem with the SuperCollider core distribution (and, in fact, not even a bug in the extension class, because it already provides a way to do what you want). Further questions should be addressed to the user forum scsynth.org or mailing list.

**Comment by jamshark70** [comment]
> I did the low notes first  4300 notes : take 2 minutes  . Then I didthe higher notes  9660  : it took  12 minutes.

This example shows that adding 10000 notes takes less than 0.3 seconds, roughly 2400 times faster than 12 minutes.

```
(
m = SimpleMIDIFile("~/nope.mid");
m.init1(1, 120, "4/4");

bench {
	10000.do {
		// please note "sort: false" here!
		// omitting this is the whole cause of your problem
		m.addNote(127.rand, 127.rand, 1000.0.rand, 1, 0, 0, 0, sort: false);
	};
	m.sortMIDIEvents;
};
)

time to run: 0.28649745300004 seconds.
time to run: 0.28392937700005 seconds.
time to run: 0.29976549100002 seconds.
```

I had mentioned the `sort` flag earlier, but if you are still seeing bad performance, I guess you didn't use that tip. It's rather crucial: if `true`, then it will sort the events for every `add`, and this will waste massive amounts of time. If `false`, then it adds without regard to order, and you can sort the events later -- *once* and once only.

SimpleMIDIFile's documentation on this point is poor. But, there is no problem with SC's code here.

> I realize that sorting it's not a linear process.

`sort`'s performance is nearly linear, actually:

```
f = { |n|
	var array = Array.fill(n, { 1000.0.rand });
	bench { array.sort };
};

f.(1000);
time to run: 0.011666774999867 seconds.

f.(10000);
time to run: 0.10721704899993 seconds.

f.(100000);
time to run: 1.145166904 seconds.

f.(1000000);
time to run: 11.485232833 seconds.
```

**Comment by brianlheim** [comment]
sorting in the best case is linearithmic (O(n lg n)) unless you are doing a very special case of sorting that SC doesn't bother with.

**Comment by jamshark70** [comment]
Ok, I did overstate the conclusion -- I do realize (of course) that four data points are not sufficient to describe the performance in general, but I didn't say that -- but the main point is still that SC's sorting performance does not degrade exponentially as claimed.

Probably enough for this thread by now.

**Comment by qcmv22** [comment]
get it...




(




 m = SimpleMIDIFile( "~/Desktop/Test.mid") ;




m.init1(2, 60, "4/4") ;  




m.addNote(58, 64, 24.0, 0.97, 127, track: 1 ,sort: false) ;

m.addNote(59, 64, 25.0, 0.97, 127, track: 1 ,sort: false) ;

m.addNote(61, 64, 26.0, 0.97, 127, track: 1 ,sort: false) ;




(...)




)




I though  ->    m.init1(2, 60, "4/4" , sort: false)        was the way to  change the default  sort == true    in addNote   










and then run this command

m.sortMIDIEvents;










With your code (1000 data)   I got around the same result ;






time to run: 0.29290978699999 seconds.


-> 0.29290978699999







Best







from  Steven S Skiena  some though about sorting :







Sorting is the basic building block that many other algorithms are built around. By understanding 

sorting, we obtain an amazing amount of power to solve other problems.




Most of the interesting ideas used in the design of algorithms appear in the context of sorting,

such as divide-and-conquer, data structures, and randomized algorithms.




Computers have historically spent more time sorting than doing anything else. 

A quarter of all mainframe cycles were spent sorting data. 

Sorting remains the most ubiquitous combinatorial algorithm problem in practice.




Sorting is the most throughly studied problem in computer science.

Literally dozens of different algorithms are known, 
most of which possess some particular advantage over all other algorithms in certain situations.


    Le mercredi 11 décembre 2019 18:30:32 UTC−5, H. James Harkins <notifications@github.com> a écrit :  
 
 
Ok, I did overstate the conclusion -- I do realize (of course) that four data points are not sufficient to describe the performance in general, but I didn't say that -- but the main point is still that SC's sorting performance does not degrade exponentially as claimed.

Probably enough for this thread by now.
